tools/pv_tree_generator/build_tree.py

```diff
@@ -1,243 +1,329 @@
+                    set(self.pos.cprops) < set(c_pos.cprops)):
+                newProp = (set(c_pos.cprops) - set(self.pos.
+                                                   cprops)).pop()
+                child_pos.append((newProp, c_pos))
+        return child_pos
+
+    def json_blob(self):
+        """generate the json blob for the property node"""
+        if self.pos.name:
+            name = self.pos.name
+        else:
+            name = self.prop
+        result = {
+            'populationType': self.pos.pop_type,
+            # used for top level reorgnize only
+            'mprop': self.pos.obs_props[0].mprop,
+            'l': text_format.format_title(name),
+            't': 'p',
+            'c': 0,
+            'cd': [],
+            'sv_set': set(),
+        }
+        return result
+
+
+class ValueNode:
+    """ Represent a value node in the PV tree
+        The value node can be
+            - a real value node representing a statsVar
+            - the node of a vertical, such as the node "Demographics"
+            - a psudo value node of super enums, such as the node 
+              "violent" under crimeType
+    """
 
-MAX_LEVEL = 6
-
-
-def build_tree_recursive(pos, level, pop_obs_spec, stat_vars,
-                         parent=None):
-    """Recursively build the ui tree"""
-    # get the property of the ui node
-    if parent:
-        property_diff = (set(pos.properties) - set(parent.pop_obs_spec.
-                                                   properties)).pop()
-        parent_pv = parent.pv
-    else:
-        property_diff = pos.properties[0]  # This is single pv pos
-        parent_pv = {}
-
-    prop_ui_node = util.UiNode(pos, parent_pv, True, property_diff)
-    result = {
-        'populationType': prop_ui_node.pop_type,
-        'l': text_format.format_title(prop_ui_node.text),
-        't': 'p',
-        'c': 0,
-        'cd': [],
-        'sv_set': set(),
-    }
-
-    # get the child specs of the current node
-    child_pos = []
-    for c_pos in pop_obs_spec[level+1]:
-        if (pos.pop_type == c_pos.pop_type and
-                set(pos.properties) < set(c_pos.properties)):
-            child_pos.append(c_pos)
-
-    child_prop_vals = []
-    for sv in stat_vars[pos.key]:
-        if sv.match_ui_node(prop_ui_node) and sv.pv[property_diff]:
-            if sv.pv[property_diff] not in child_prop_vals:
-                # if this is the first statsvar of this node, build the node
-                # and corresponding child nodes
-                child_prop_vals.append(sv.pv[property_diff])
-                value_ui_pv = collections.OrderedDict()
-                for prop, val in parent_pv.items():
-                    value_ui_pv[prop] = val
-                value_ui_pv[property_diff] = sv.pv[property_diff]
-                value_ui_node = util.UiNode(
-                    pos, value_ui_pv, False, property_diff)
+    def __init__(self, value: str, leafs: List['valueLeaf'], 
```

Comment:
should this be List[valueLeaf] (and actually better to be List[ValueLeaf]?

Reply:
Changed to List[valueLeaf].
FYI, I use List['ValueNode'] in class PropertyNode, because to use List[ValueNode], class ValueNode has be defined before PropertyNode.